### PR TITLE
Add support for CreationTime and File Attributes in Windows

### LINKED ISSUE #2075: Directory and file attribute "hidden" not restored
Output of `restic version`
--------------------------
restic 0.9.3 compiled with go1.11.1 on windows/amd64
running in Windows 10

How did you run restic exactly?
-------------------------------
.\restic restore --repo D:\test\ --target '/restore' latest --host XXX --password-file password.txt 

No environmental variables were used in this test.

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
Local file system - no backend server

Expected behavior
-----------------
All files and directories are restored along with their attribute settings. Specifically - the hidden attribute.

Actual behavior
---------------

All files and directories were restored perfectly except the hidden attribute was lost.  For example a .git directory will be fully restored but was visible in the restore. This behavior is consistent across multiple scenarios.

Steps to reproduce the behavior
-------------------------------
Create a Windows file and directory marked hidden. Backup and restore then and the hidden attribute is unset.

Do you have any idea what may have caused this?
-----------------------------------------------
No

Do you have an idea how to solve the issue?
-------------------------------------------
I cannot tell if the hidden attribute is lost in the backup phase or the restore phase.

Did restic help you or made you happy in any way?
-------------------------------------------------
I think restic is a great tool.

### PR DESCRIPTION
**Pull request by mchowdhury**
What does this PR change? What problem does it solve?
-----------------------------------------------------
File created time (an NTFS specific attribute) and File Attributes don't get backed up or restored on Windows backups. See issue #3478 and #2075

Since we don't want to modify ctimes (because it's used for file change comparison #2212) we store creation time as an extended attribute and restore the creation time value when the file is restored.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

This issue was filed in #3478 and #2075
A related issue is #1401 for backing up entire NTFS extended attributes. This uses microsoft provided go-winio library. In the future this would be a better approach but would be a significant re-write. See this comment: https://github.com/restic/restic/issues/1401#issuecomment-392959322

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added tests for all code changes.
- [ ] I have added documentation for relevant changes (in the manual).
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [ ] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

### CODE REVIEW

**Comment by DRON-666** [comment]
Can you add windows attributes as well? This would solve issue #2075.

**Comment by mchowdhury** [comment]
> Can you add windows attributes as well? This would solve issue #2075.

Ideally it would be nice to have all the missing attributes included. This can be done what's described in https://github.com/restic/restic/issues/1401#issuecomment-392959322

It's more effort but can be done. I would probably need some assistance with the implementation.

**Comment by DRON-666** [comment]
>Ideally it would be nice to have all the missing attributes included.

I fully agree, but adding standard attributes to you PR is just the use of `SetFileAttributes` instead of `SetFileTime`.

>This can be done what's described in #1401 (comment)

I don't think that `BackupRead/Write` API is compatible with `restic` architecture. It does not allow you to choose the type of requested information. Anyway `BackupRead/Write` is just a [userland code with bunch of NT-function calls](https://github.com/KernelPanic-OpenSource/Win2K3_NT_base/blob/master/win32/client/backup.c#L1628) and you don't need to use it.

**Comment by mchowdhury** [comment]
> I fully agree, but adding standard attributes to you PR is just the use of `SetFileAttributes` instead of `SetFileTime`.

This was a good idea. I updated and tested the PR with saving and restoring file attributes in addition to creation time

File: internal/restic/node_windows.go [hunk 1]
```diff
@@ -22,18 +27,104 @@ func (node Node) restoreSymlinkTimestamps(path string, utimes [2]syscall.Timespe
...
+
+		s, ok := fileinfo.Sys().(*syscall.Win32FileAttributeData)
+		if ok && s != nil {
+			var creationTime bytes.Buffer
+			enc := gob.NewEncoder(&creationTime)
+			
+			if err := enc.Encode(syscall.NsecToTimespec(s.CreationTime.Nanoseconds())); err != nil {
+				return nil, errors.Wrap(err, "Getxattr")
+			}
+			return creationTime.Bytes(), nil
```
Comment (DRON-666) [inline internal/restic/node_windows.go:45]: I don' think you need all this conversions and encoding. Simple `binary` conversion is enough:
```          
var creationTime [8]byte
binary.LittleEndian.PutUint32(creationTime[0:4], s.CreationTime.LowDateTime)
binary.LittleEndian.PutUint32(creationTime[4:8], s.CreationTime.HighDateTime)
return creationTime[:], nil
```

File: internal/restic/node_windows.go [hunk 2]
```diff
@@ -22,18 +27,104 @@ func (node Node) restoreSymlinkTimestamps(path string, utimes [2]syscall.Timespe
...
+	
+		var creationTime syscall.Timespec
+		   dec := gob.NewDecoder(&inputData)
+		
+		if err := dec.Decode(&creationTime); err != nil {
+			return errors.Wrap(err, "Setxattr")
+		}
+		   
+		c := syscall.NsecToFiletime(time.Unix(creationTime.Unix()).UnixNano())
+		if err := syscall.SetFileTime(h, &c, nil, nil); err != nil {
```
Comment (DRON-666) [inline internal/restic/node_windows.go:110]: Same as above:
```
var creationTime syscall.Filetime
creationTime.LowDateTime = binary.LittleEndian.Uint32(data[0:7])
creationTime.HighDateTime = binary.LittleEndian.Uint32(data[4:8])
if err := syscall.SetFileTime(h, &creationTime, nil, nil); err != nil {
```

File: internal/restic/node_windows.go [hunk 3]
```diff
@@ -22,18 +27,104 @@ func (node Node) restoreSymlinkTimestamps(path string, utimes [2]syscall.Timespe
...
+			return errors.Wrap(err, "Setxattr")
+		}
+		return nil
+	case "FileAttributes":
+		var attrs uint32
+		inputData := bytes.NewBuffer(data)
+		
+		if err := binary.Read(inputData, binary.LittleEndian, &attrs); err != nil {
+			return errors.Wrap(err, "Setxattr")
+		}
```
Comment (DRON-666) [inline internal/restic/node_windows.go:120]: And here too:
```
attrs := binary.LittleEndian.Uint32(data)
```

**Comment by mchowdhury** [comment]
@DRON-666 Thanks again for the feedback. I made the necessary changes and tested them.
Let me know if you have any other suggestions.

**Comment by mchowdhury** [comment]
@DRON-666 What else needs to be done to get this merged?

**Comment by hgraeber** [comment]
It would be nice to add security information the same. Would mean to use GetFileSecurity and SetFileSecurity the get and set the DACL. Converting the descriptor to/from SDDL string representation would be even better. ConvertSecurityDescriptorToStringSecurityDescriptor and ConvertSecurityDescriptorFromStringSecurityDescriptor are helpful for that.

Maybe you consider to use a readable representation for the other attributes, too.

**Comment by mchowdhury** [comment]
> It would be nice to add security information the same. Would mean to use GetFileSecurity and SetFileSecurity the get and set the DACL. Converting the descriptor to/from SDDL string representation would be even better. ConvertSecurityDescriptorToStringSecurityDescriptor and ConvertSecurityDescriptorFromStringSecurityDescriptor are helpful for that.
> 
> Maybe you consider to use a readable representation for the other attributes, too.

@hgraeber GetFileSecurity and SetFileSecurity are not currently supported by the standard go syscall library. I'm open to other suggestions. Otherwise, should we make a separate issue for file security attributes and get this merged?

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
Mapping the creationTime / fileAttributes in to xattrs feels wrong, please just add new (optional) fields to the `Node` struct. I'd also prefer to store the creationTime as a regular timestamp and not some Windows specific timestamp format. Btw, the creationTime is known as birthTime on Linux, so it might be more appropriate to use that name (or short `btime` for the json attribute). Even though we can't write that timestamp on Linux.

The FileAttributes should also be accessible from the struct returned at https://github.com/restic/restic/blob/cccc17e4e91facdaf710dc82a59653c5c0de3156/internal/restic/node_windows.go#L56 . Adding two redundant syscalls (the two os.Stat) calls will likely hurt performance.

Judging from https://learn.microsoft.com/en-us/windows/win32/fileio/file-attribute-constants I'm not sure whether it's a good idea to blindly store and restore arbitrary file attributes. But I'm not familiar enough with the Windows API to accurately judge that.

Restoring access permissions etc. definitely doesn't belong into this PR.

**Comment by mchowdhury** [comment]
@MichaelEischer I can move the creation time directly to `ctim()` since we already make the fileattributes syscall in the struct.


For Windows file attributes, these are essentially the same as Linux extended attributes and should be backed up and restored. I figured adding it here would be the cleanest way.

How do you propose we integrate and restore them if they are added as optional attributes to the node struct?

**Comment by hgraeber** [comment]
> Judging from https://learn.microsoft.com/en-us/windows/win32/fileio/file-attribute-constants I'm not sure whether it's a good idea to blindly store and restore arbitrary file attributes. 

I think the file attributes shall not be copied blindly, because they may have some side effect and cannot be simply restored. (https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-setfileattributesa). It is necessary to set flags on file creation or using special functions to reach the desired effect.

**Comment by hgraeber** [comment]
@mchowdhury

> GetFileSecurity and SetFileSecurity are not currently supported by the standard go syscall library. I'm open to other suggestions. Otherwise, should we make a separate issue for file security attributes and get this merged?

Agreed. I will write an issue for this. There we can discuss how to implement backup of windows file security information.

**Comment by mchowdhury** [comment]
@hgraeber I don't quite understand the approach you are describing, and unsure how to implement it without impacting other platforms (node.go). Should creationTime follow the same approach? If not, we can separate file attributes to a separate PR and just focus on creationTime.

Would be happy to have your help on the implementation if you have bandwidth.

**Comment by ProactiveServices** [comment]
Have you considered the more unusual attributes and what happens if they cannot be restored, or what might unexpected behaviour might occur if they are restored? There's a host of weird and wonderful attributes other than the usual hidden/read-only/system. How is the restoration of a sparse file handled? What about attributes that only kernel-level code can restore? Will a restore apply to a file marked as temporary? How are you handling EFS-encrypted data? Files that have attributes that are not supported on the target filesystem?

**Comment by mchowdhury** [comment]
@MichaelEischer After more research, I like your suggestion with `btime` adding it to the node, also since darwin also supports btime. I will make a new PR to support creationTime on windows and os x.  

Let's leave this PR open to discuss File Attributes on Windows. File attributes got pulled into this PR from @DRON-666 suggestion to fix issue https://github.com/restic/restic/issues/2075.

**Comment by heapxor** [comment]
hello,
is that merge request stable and tested? thanks

**Comment by hanacchi** [comment]
any progress on getting this merged? i would love to use restic but lack of creation time support on windows is holding me back

**Comment by Dobatymo** [comment]
A file copy using the Windows Explorer doesn't preserve file creation dates (only modification dates). So I don't think restic should restore them either.

**Comment by tszynalski** [comment]
> A file copy using the Windows Explorer doesn't preserve file creation dates (only modification dates). So I am not sure restic should restore them either.

Copying a file is not the same as restoring a file from backup. When you copy a file, you create a second file with the same contents – but it's not the _same_ file, so the creation date should be different. When you back up a file, you want to be able to "turn back time" and go back _to the way things were_ in the event of failure. If something is lost (such as a piece of metadata), then the restore was not complete.

**Comment by MichaelEischer** [comment]
After taking a look at https://github.com/restic/restic/pull/4611 , I think we should follow that approach. The `GenericAttributes` suggested there just look like what I had in mind.

Sorry for not finding time to discuss this PR!

**Pull request closed without merging**